cpp/src_prims/matrix/batched_matrix.hpp

```diff
@@ -0,0 +1,440 @@
+  }
+
+  auto num_batches = A.batches();
+  int m = A.shape().first;
+  int n = A.shape().second;
+
+  BatchedMatrix C(m, n, num_batches, A.cublasHandle(), A.allocator(),
+                  A.stream());
+
+  LinAlg::binaryOp(C[0], A[0], B[0], m * n * num_batches, binary_op,
+                   A.stream());
+
+  return C;
+}
+
+/* Multiplies each matrix in a batch-A with it's batch-B counterpart.
+ * A = [A1,A2,A3], B=[B1,B2,B3]
+ * return [A1*B1, A2*B2, A3*B3]
+ * @param A Matrix `A`
+ * @param B Matrix `B`
+ * @returns The result of the batched matrix-matrix multiplication of `A` x `B`
+ */
+BatchedMatrix operator*(const BatchedMatrix& A, const BatchedMatrix& B) {
+  return b_gemm(A, B);
+}
+
+/* Adds two batched matrices together element-wise.
+ * @param Matrix A
+ * @param Matrix B
+ * @return A+B
+ */
+BatchedMatrix operator+(const BatchedMatrix& A, const BatchedMatrix& B) {
+  return b_aA_op_B(A, B, [] __device__(double a, double b) { return a + b; });
+}
+
+/* Subtract two batched matrices together element-wise.
+ * @param Matrix A
+ * @param Matrix B
+ * @return A-B
+ */
+BatchedMatrix operator-(const BatchedMatrix& A, const BatchedMatrix& B) {
+  return b_aA_op_B(A, B, [] __device__(double a, double b) { return a - b; });
+}
+
+/* Solve Ax = b for given batched matrix A and batched vector b
+ * @param Matrix A
+ * @param vector b
+ * @return A\b
+ */
+BatchedMatrix b_solve(const BatchedMatrix& A, const BatchedMatrix& b) {
+  auto num_batches = A.batches();
+  const auto& handle = A.cublasHandle();
+
+  int n = A.shape().first;
+  auto allocator = A.allocator();
+  int* P = (int*)allocator->allocate(sizeof(int) * n * num_batches, 0);
+  int* info = (int*)allocator->allocate(sizeof(int) * num_batches, 0);
+
+  BatchedMatrix Ainv(n, n, num_batches, A.cublasHandle(), A.allocator(),
+                     A.stream());
+
+  CUBLAS_CHECK(
+    cublasDgetrfBatched(handle, n, A.data(), n, P, info, num_batches));
+  CUBLAS_CHECK(cublasDgetriBatched(handle, n, A.data(), n, P, Ainv.data(), n,
+                                   info, num_batches));
+
+  BatchedMatrix x = Ainv * b;
+
+  allocator->deallocate(P, sizeof(int) * n * num_batches, A.stream());
+  allocator->deallocate(info, sizeof(int) * num_batches, A.stream());
+
+  return x;
+}
+
+/* The batched kroneker product A (x) B for given batched matrix A and batched matrix B
+ * @param Matrix A
+ * @param Matrix B
+ * @return A (x) B
+ */
+BatchedMatrix b_kron(const BatchedMatrix& A, const BatchedMatrix& B) {
+  int m = A.shape().first;
+  int n = A.shape().second;
+
+  int p = A.shape().first;
+  int q = B.shape().second;
+
+  // resulting shape
+  int km = m * p;
+  int kn = n * q;
+
+  BatchedMatrix AkB(km, kn, A.batches(), A.cublasHandle(), A.allocator(),
+                    A.stream());
+
+  // run kronecker...
+  dim3 threads(std::min(p, 32), std::min(q, 32));
+  kronecker_product_kernel<<<A.batches(), threads, 0, A.stream()>>>(
+    A.data(), m, n, B.data(), p, q, AkB.data(), km, kn);
+
```

Comment:
it is good to put a `CUDA_CHECK(cudaGetLastError());` here just to catch any issues with launching the kernel itself. (else becomes hard to pin-point the location of error, since these launches are host-asynchronous)